contrib/babelfishpg_common/src/datetimeoffset.c

```diff
@@ -832,3 +836,86 @@ EncodeDatetimeoffsetTimezone(char *str, int tz, int style)
 
 	*tmp = '\0';
 }
+
+Datum
+dateadd_datetimeoffset(PG_FUNCTION_ARGS) {
+	text    *field     = PG_GETARG_TEXT_PP(0);
+	int      num       = PG_GETARG_INT32(1);
+	tsql_datetimeoffset *d = PG_GETARG_DATETIMEOFFSET(2);
+
+	char	   *lowunits;
+	int			type,
+				val;
+	tsql_datetimeoffset *result;
+	Interval   *interval;
+
+	lowunits = downcase_truncate_identifier(VARDATA_ANY(field),
+									VARSIZE_ANY_EXHDR(field),
+									false);
+
+	type = DecodeUnits(0, lowunits, &val);
+
+	if(strncmp(lowunits, "doy", 3) == 0 || strncmp(lowunits, "dayofyear", 9) == 0) {
+		type = UNITS;
+		val = DTK_DOY;
+	}
+
+	if(strncmp(lowunits, "nanosecond", 11) == 0) {
+		type = UNITS;
+		val = DTK_NANO;
+	}
+	if(strncmp(lowunits, "weekday", 7) == 0) {
+		type = UNITS;
+		val = DTK_DAY;
+	}
+
+	if(type == UNITS) {
+		switch(val) {
+			case DTK_YEAR:
+				interval = (Interval *) DirectFunctionCall7(make_interval, num, 0, 0, 0, 0, 0, 0);
+				break;
+			case DTK_QUARTER:
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, num * 3, 0, 0, 0, 0, 0);
+				break;
+			case DTK_MONTH:
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, num, 0, 0, 0, 0, 0);
+				break;
+			case DTK_WEEK:
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, num, 0, 0, 0, 0);
+				break;
+			case DTK_DAY:
+			case DTK_DOY:
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, num, 0, 0, 0);
+				break;
+			case DTK_HOUR:
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, num, 0, 0);
+				break;
+			case DTK_MINUTE:
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, num, 0);
+				break;
+			case DTK_SECOND:
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, 0, Float8GetDatum(num));
+				break;
+			case DTK_MILLISEC:
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, 0, Float8GetDatum((float) num * 0.001));
+				break;
+			case DTK_MICROSEC:
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, 0, Float8GetDatum((float) num * 0.000001));
+				break;
+			case DTK_NANO:
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, 0, Float8GetDatum((float) num * 0.000000001));
+				break;
+			default:
+				elog(ERROR, "the datepart %s is not supported by function dateadd for datatype datetimeoffset",
+				 	lowunits);
+				break;
+		}
+	} else {
+		elog(ERROR, "the datepart %s is not supported by function dateadd for datatype datetimeoffset", lowunits);
```

Comment:
The original msg on the script impl is : 

		RAISE EXCEPTION '"%" is not a recognized dateadd option.', datepart;